chacra/async/recurring.py

```diff
@@ -84,21 +84,25 @@ def purge_repos(_now=None):
         project_repos = models.Repo.filter_by(project=p)
 
         # get the configuration for current project
-        ref_names = purge_rotation[project_name].keys()
         purge_project = purge_rotation.get(project_name)
-
-        for ref_name in ref_names:
-            purge_ref = purge_project[ref_name]
-            lifespan = now - datetime.timedelta(days=purge_ref.get('days', 14))
-            keep_minimum = purge_ref.get('keep_minimum', 0)
-            ref_repos = project_repos.filter_by(ref=ref_name)
-
-            # filter the query further, with the offset being the
-            # minimum to keep, and lifespan how old they should be
-            repos = ref_repos.filter(
-                    models.Repo.modified < lifespan).order_by(
-                        desc(models.Repo.modified)).offset(keep_minimum).all()
-            delete_repositories(repos, lifespan, keep_minimum)
+        # can be ref and/or flavor
+        for proj_filter, repo_attrs in purge_project.items():
+            for attr in repo_attrs:
+                purge_attr = repo_attrs[attr]
+                lifespan = now - datetime.timedelta(days=purge_attr.get('days', 14))
+                keep_minimum = purge_attr.get('keep_minimum', 0)
+                # add similar conditional to filter_by other repo attributes
+                if proj_filter == 'ref':
+                    attr_repos = project_repos.filter_by(ref=attr)
+                elif proj_filter == 'flavor':
```

Comment:
The elif is fine here because proj_filter has only one value per iteration and can only be 'ref' or 'flavor' .

The code is muddled and hard to read, but here is an example of the layer in the purge_rotation proj_filter is assigned to. https://github.com/ceph/chacra/blob/e474567b85b921564fdb1b371a81f22fd3746f78/deploy/playbooks/roles/common/templates/prod.py.j2#L143

You are right, we want to allow for filtering for both. There could be a project that has both "ref" and "flavor" in it's purge rotation, they would each be a different iteration here (https://github.com/ceph/chacra/blob/e474567b85b921564fdb1b371a81f22fd3746f78/chacra/async/recurring.py#L89) and I have a test that tests for that.